### PR TITLE
Move Transaction to ES6 class syntax

### PR DESCRIPTION
**Pull request by junderw**
Edit: Decided to only do the syntax change here, getters and setters etc. should be in a different PR.

Two goals:

1. ES6 usage all throughout Transaction. Including class constructor.
2. change getHash, getId etc. to tx.hash tx.id etc.

One question:

block.js L153 typeforce: What should I change it to?

### CODE REVIEW

**Comment by dcousens** [comment]
> block.js no L153 no typeforce: What should I change it to?

I thought issues like this might prop,  esp. around `getHash` etc.

https://github.com/bitcoinjs/bitcoinjs-lib/blob/0d76f00917d8654377b4434ef94762fcef2402c1/src/block.js#L152-L161

`(t) => t.hash` would simply replace the inner-loop.  If we type-check it,  we double-down the on the cost.
It would have to either be a constant structure that we can lazily evaluate to be sane on repeated `.hash` access (like we do in `ECPair`),  or we should probably leave `getHash` as is.

`.id` falls to the same thing.

A constant structure approach would break _a lot_.

**Comment by dcousens** [comment]
d85c4c9 ... so many dirty flags :cry:

**Comment by dcousens** [comment]
@junderw we can still do this,  but you need to drop the

> change getHash, getId etc. to tx.hash tx.id etc.

changes.

**Comment by junderw** [comment]
Hmm, to be honest, rebasing and re-thinking is kind of hairy, so I am thinking of just closing this and starting all over with a different approach maybe...?

**Comment by dcousens** [comment]
OK,  if you think it'd be easier.

**Comment by junderw** [comment]
Decided to only do Syntax for now, much easier. unit tests and standard tests passed locally.

File: src/transaction.js [hunk 1]
```diff
@@ -7,35 +7,20 @@ const typeforce = require('typeforce')
 const types = require('./types')
 const varuint = require('varuint-bitcoin')
 
-function varSliceSize (someScript) {
+const varSliceSize = (someScript) => {
   const length = someScript.length
```
Comment (dcousens) [inline src/transaction.js]: why?

File: src/transaction.js [hunk 2]
```diff
@@ -7,35 +7,20 @@ const typeforce = require('typeforce')
...
 
   return varuint.encodingLength(length) + length
 }
 
-function vectorSize (someVector) {
+const vectorSize = (someVector) => {
   const length = someVector.length
 
-  return varuint.encodingLength(length) + someVector.reduce(function (sum, witness) {
```
Comment (dcousens) [inline src/transaction.js]: OK, I do prefer the shorter style here,  but,  only as a matter of consistency

File: src/transaction.js [hunk 3]
```diff
@@ -7,35 +7,20 @@ const typeforce = require('typeforce')
...
 const varuint = require('varuint-bitcoin')
 
-function varSliceSize (someScript) {
+const varSliceSize = (someScript) => {
   const length = someScript.length
 
   return varuint.encodingLength(length) + length
 }
 
-function vectorSize (someVector) {
```
Comment (dcousens) [inline src/transaction.js]: this seems silly,  searching for `function vectorSize` is immensely useful compared to the above syntax, which, although constant, makes me think of this `function` as somehow variable...

File: src/transaction.js [hunk 4]
```diff
@@ -46,447 +31,525 @@ const BLANK_OUTPUT = {
...
+    this.ins = []
+    this.outs = []
   }
 
-  function readUInt32 () {
-    const i = buffer.readUInt32LE(offset)
-    offset += 4
-    return i
+  static get DEFAULT_SEQUENCE () {
+    return 0xffffffff
```
Comment (dcousens) [inline src/transaction.js]: is this _the_ way to do constants?

File: src/transaction.js (hunk 1 above)
Comment (junderw) [reply src/transaction.js]: Personal preference. hoisted functions make me feel dirty inside nowadays... I'm not religious about it tho.

File: src/transaction.js (hunk 3 above)
Comment (junderw) [reply src/transaction.js]: I love regex, so I'd just search `const xxxxx = \S+ => ` where xxxxx is the function I'm looking for... in fact that query is in my favorites actually...

Again, preference, and I'm not too set on it.

File: src/transaction.js (hunk 4 above)
Comment (junderw) [reply src/transaction.js]: This is the best way I found. Open to other suggestions. (I'm guessing you want one constant per line?)

**Comment by junderw** [comment]
You can add a commit tweaking the preference stuff if you'd like, I'm not too set on it either way.

File: src/transaction.js (hunk 1 above)
Comment (dcousens) [reply src/transaction.js]: for now, lets leave this as it was,  and then we can discuss the new style separately?

**Comment by junderw** [comment]
Reverted

**Comment by dcousens** [comment]
LGTM, shame the diff is a mess.

I wonder if we can force it to 'align'?

**Comment by junderw** [comment]
idk lol... refactors are usually messy...

**Comment by junderw** [comment]
Forgot to merge this lol

**Comment by dcousens** [comment]
Can we revert this and merge it into an `es6` branch,  then we can retarget each PR to that branch

**Pull request merged**